packages/devtools_app/lib/src/memory/flutter/memory_filter.dart

```diff
@@ -0,0 +1,391 @@
+class FilteredLibraries {
+  final List<String> _filteredLibraries = [
+    _dartLibraryUriPrefix,
+    _collectionLibraryUri,
+    _flutterLibraryUriPrefix,
+    _intlLibraryUri,
+    _vectorMathLibraryUri,
+  ];
+
+  static String normalizeLibraryUri(Library library) {
+    final uriParts = library.uri.split('/');
+    final firstPart = uriParts.first;
+    if (firstPart.startsWith(_dartLibraryUriPrefix)) {
+      return _dartLibraryUriPrefix;
+    } else if (firstPart.startsWith(_flutterLibraryUriPrefix)) {
+      return _flutterLibraryUriPrefix;
+    } else {
+      return firstPart;
+    }
+  }
+
+  List<String> get librariesFiltered =>
+      _filteredLibraries.toList(growable: false);
+
+  bool get isDartLibraryFiltered =>
+      _filteredLibraries.contains(_dartLibraryUriPrefix);
+
+  bool get isFlutterLibraryFiltered =>
+      _filteredLibraries.contains(_flutterLibraryUriPrefix);
+
+  void clearFilters() {
+    _filteredLibraries.clear();
+  }
+
+  void addFilter(String libraryUri) {
+    _filteredLibraries.add(libraryUri);
+  }
+
+  void removeFilter(String libraryUri) {
+    _filteredLibraries.remove(libraryUri);
+  }
+
+  // Keys in the libraries map is a normalized library name.
+  List<String> sort() => _filteredLibraries..sort();
+
+  bool isDartLibrary(Library library) =>
+      library.uri.startsWith(_dartLibraryUriPrefix);
+
+  bool isFlutterLibrary(Library library) =>
+      library.uri.startsWith(_flutterLibraryUriPrefix);
+
+  bool isDartLibraryName(String libraryName) =>
+      libraryName.startsWith(_dartLibraryUriPrefix);
+
+  bool isFlutterLibraryName(String libraryName) =>
+      libraryName.startsWith(_flutterLibraryUriPrefix);
+
+  bool isLibraryFiltered(String libraryName) =>
+      // Are dart:* libraries filtered and its a Dart library?
+      (_filteredLibraries.contains(_dartLibraryUriPrefix) &&
+          isDartLibraryName(libraryName)) ||
+      // Are package:flutter* filtered and its a Flutter package?
+      (_filteredLibraries.contains(_flutterLibraryUriPrefix) &&
+          isFlutterLibraryName(libraryName)) ||
+      // Is this library filtered?
+      _filteredLibraries.contains(libraryName);
+}
+
+/// State of the libraries and packages (hidden or not) for the filter dialog.
+class LibraryFilter {
+  LibraryFilter(this.normalized, this.hide);
+
+  /// Displayed library name.
+  final String normalized;
+
+  /// Classes in this library hidden.
+  bool hide;
+}
+
+class SnapshotFilterDialog extends StatefulWidget {
+  const SnapshotFilterDialog(this.controller);
+
+  final MemoryController controller;
+
+  @override
+  SnapshotFilterState createState() => SnapshotFilterState();
+}
+
+/// A FlatButton used to close a containing dialog - Cancel.
+class DialogCancelButton extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return FlatButton(
+      onPressed: () {
+        Navigator.of(context, rootNavigator: false).pop('dialog');
```

Comment:
can you make 'dialog' a constant?
Also promote these widgets to a shared ui component folder in devtools as they seem to apply to all dialogs we would want in the future as long as we use consistent conventions.

Reply:
Done.